vector/src/main/java/im/vector/app/features/userdirectory/KnownUsersFragment.kt

```diff
@@ -107,6 +120,47 @@ class KnownUsersFragment @Inject constructor(
         }
     }
 
+    private fun setupAddByQrCodeView() {
+        val qrStartForActivityResult = registerStartForActivityResult { activityResult ->
+            if (activityResult.resultCode == Activity.RESULT_OK) {
+                val result = QrCodeScannerActivity.getResultText(activityResult.data)!!
+
+                // Latter condition to be replaced by URL(result).protocol == "matrix" when MSC2312 is complete
+                if (result.isValidUrl() && URL(result).host == "matrix.to") {
+                    // Remove query string parameters and initial slash from URL().ref to get MXID
+                    val mxid = URL(result).ref.split("?")[0].drop(1)
+
+                    val existingDm = session.getExistingDirectRoomWithUser(mxid)
+
+                    if (existingDm == null) {
+                        // The following assumes MXIDs are case insensitive. Change if needed
+                        if (mxid.toLowerCase(Locale.getDefault()) != session.myUserId.toLowerCase(Locale.getDefault())) {
+                            // Try to get user from known users and fall back to creating a User object from MXID
+                            val qrInvitee = if (session.getUser(mxid) != null) session.getUser(mxid)!! else User(mxid, null, null)
+
+                            createDirectRoomViewModel.handle(
+                                    CreateDirectRoomAction.CreateRoomAndInviteSelectedUsers(setOf(PendingInvitee.UserPendingInvitee(qrInvitee)))
+                            )
+                        } else {
+                            Toast.makeText(requireContext(), R.string.dm_self, Toast.LENGTH_SHORT).show()
+                        }
+                    } else {
+                        navigator.openRoom(requireContext(), existingDm.roomId, null, false)
+                        Toast.makeText(requireContext(), R.string.dm_already_exists, Toast.LENGTH_SHORT).show()
+                        requireActivity().finish()
+                    }
+                } else {
+                    Toast.makeText(requireContext(), R.string.invalid_qr_code_uri, Toast.LENGTH_SHORT).show()
+                }
+            } else {
+                Toast.makeText(requireContext(), R.string.qr_code_not_scanned, Toast.LENGTH_SHORT).show()
+            }
+        }
+        addByQrCode.debouncedClicks {
+            QrCodeScannerActivity.startForResult(requireActivity(), qrStartForActivityResult)
```

Comment:
You have to request the permission to use the camera first. See how it is done in `VerificationChooseMethodFragment`. To test, you can deny the camera permission in the system setting of the application.